integration/kotlinx-coroutines-play-services/src/Tasks.kt

```diff
@@ -44,45 +37,97 @@ public fun <T> Deferred<T>.asTask(): Task<T> {
-            CompletableDeferred<T>().apply { if (isCanceled) cancel() else complete(result as T) }
+        if (e == null) {
+            if (isCanceled) {
+                deferred.cancel()
+            } else {
+                @Suppress("UNCHECKED_CAST")
+                deferred.complete(this.result as T)
+            }
         } else {
-            CompletableDeferred<T>().apply { completeExceptionally(e) }
+            deferred.completeExceptionally(e)
+        }
+    } else if (cancellationTokenSource.token.isCancellationRequested) {
+        // The task hasn't completed, yet cancellation was already requested.
+        // Interpret this by cancelling immediately (no way to cancel the task)
+        deferred.cancel()
+    } else {
+        addOnCompleteListener {
+            val e = it.exception
+            if (e == null) {
+                @Suppress("UNCHECKED_CAST")
+                if (isCanceled) deferred.cancel() else deferred.complete(it.result as T)
+            } else {
+                deferred.completeExceptionally(e)
+            }
+        }
+
+        cancellationTokenSource.token.onCanceledRequested {
+            deferred.cancel()
         }
     }
 
-    val result = CompletableDeferred<T>()
-    addOnCompleteListener {
-        val e = it.exception
-        if (e == null) {
-            @Suppress("UNCHECKED_CAST")
-            if (isCanceled) result.cancel() else result.complete(it.result as T)
-        } else {
-            result.completeExceptionally(e)
+    deferred.invokeOnCompletion {
+        if (it is CancellationException) {
+            cancellationTokenSource.cancel()
         }
     }
-    return result
+
+    return deferred
 }
 
 /**
  * Awaits for completion of the task without blocking a thread.
  *
+ * Prefer passing the corresponding [CancellationTokenSource] if the [Task] can be created with a [CancellationToken]
+ * to support bi-directional cancellation.
+ *
  * This suspending function is cancellable.
  * If the [Job] of the current coroutine is cancelled or completed while this suspending function is waiting, this function
  * stops waiting for the completion stage and immediately resumes with [CancellationException].
  */
-public suspend fun <T> Task<T>.await(): T {
+public suspend fun <T> Task<T>.await(): T = await(CancellationTokenSource())
+
+/**
+ * Awaits for completion of the task with a [CancellationTokenSource] to control cancellation.
+ *
+ * This suspending function is cancellable.
+ * If the [Job] of the current coroutine is cancelled or completed while this suspending function is waiting, this function
+ * cancels the [cancellationTokenSource] and throws a [CancellationException].
+ *
+ * If the task is cancelled, then [cancellationTokenSource] will be canceled and this function will throw a
+ * [CancellationException].
+ * If the [cancellationTokenSource] is cancelled, then this function will throw a [CancellationException].
+ */
+@ExperimentalCoroutinesApi
```

Comment:
Both of these could be combined to have `cancellationTokenSource` be a default argument, but then I couldn't apply `@ExperimentalCoroutinesApi` to just the new overload.